File: src/uom/DimensionedNumber.java
/*
 * Copyright (c) 2015 - 2015 Paul Griffioen
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package uom;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;

public class DimensionedNumber {
    
    private final Unit unit;
    private final BigDecimal factor;

    public DimensionedNumber() {
        unit = new PowerProduct();
        factor = BigDecimal.ONE;
    }

    public DimensionedNumber(Unit unit) {
        this.unit = unit;
        factor = BigDecimal.ONE;
    }

    public DimensionedNumber(BigDecimal factor) {
        unit = new PowerProduct();
        this.factor = factor;
    }
        
    public DimensionedNumber(BigDecimal factor, Unit unit) {
        this.unit = unit;
        this.factor = factor;
    }

    public BigDecimal factor() {
        return factor;
    }

    public Unit unit() {
        return unit;
    }
    
    @Override
    public int hashCode() {
        return unit.hashCode();
    }

    @Override
    public boolean equals(Object other) {
        if (other == this) {
            return true;
        }
        if (!(other instanceof DimensionedNumber)) {
            return false;
        }
        DimensionedNumber otherNumber = (DimensionedNumber) other;
        if (factor.compareTo(otherNumber.factor()) != 0) {
            return false;
        }
        return unit.equals(otherNumber.unit);
    }

    @Override
    public String toString() {
        return factor.toString() + " " + unit.toString();
    }

    public DimensionedNumber multiply(BigDecimal factor) {
        return new DimensionedNumber(this.factor.multiply(factor), unit);
    }

    public DimensionedNumber multiply(DimensionedNumber other) {
        return new DimensionedNumber(other.factor.multiply(factor), other.unit.multiply(unit));
    }

    public DimensionedNumber raise(Fraction power) {
        BigDecimal raisedFactor;
        if (power.isInt()) {
            int pow = power.intValue();
            if (0 < pow) {
                raisedFactor = factor.pow(pow, MathContext.DECIMAL128);
            } else {
                raisedFactor = BigDecimal.ONE.divide(factor.pow(-pow, MathContext.DECIMAL128), 25, RoundingMode.HALF_UP);
            }
        } else {
            raisedFactor = new BigDecimal(Math.pow(factor.doubleValue(), power.doubleValue()));
        }
        return new DimensionedNumber(raisedFactor, unit.raise(power));
    }

    public DimensionedNumber reciprocal() {
        return raise(new Fraction(-1));
    }

    public DimensionedNumber flat() {
        return unit.flat().multiply(factor);
    }

    public String toText() {
        return factor.toString() + " " + unit.toText();
    }
}

File: src/pacioli-js/dimensioned-number.js
/* Runtime Support for the Pacioli language
 *
 * Copyright (c) 2015 Paul Griffioen
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

var Pacioli = Pacioli || {};

// -----------------------------------------------------------------------------
// 1. Dimensioned Numbers
//
// A DimensionedNumber pairs a number and a unit
//
// -----------------------------------------------------------------------------

Pacioli.DimensionedNumber = function (factor, unit) {
    this.factor = typeof factor === "number" ? factor : 1;
    this.unit = unit || Pacioli.unit();
}

Pacioli.DimensionedNumber.prototype.equals = function (other) {
    return this.factor === other.factor && this.unit.equals(other.unit);
}

Pacioli.DimensionedNumber.prototype.isDimensionless = function () {
    return this.unit.isDimensionless();
}

Pacioli.DimensionedNumber.prototype.flat = function () {
    var flatUnit = this.unit.flat();
    return new Pacioli.DimensionedNumber(this.factor * flatUnit.factor, flatUnit.unit);
}

Pacioli.DimensionedNumber.prototype.toText = function () {
    var unitText = this.unit.equals(Pacioli.unit()) ? "" : " " + this.unit.symbolized().toText();
    return this.factor + unitText;
}

Pacioli.DimensionedNumber.prototype.toFixed = function (n) {
    var unitText = this.unit.equals(Pacioli.unit()) ? "" : " " + this.unit.symbolized().toText();
    return this.factor.toFixed(n) + unitText;
}

Pacioli.DimensionedNumber.prototype.toDOM = function () {
    var fragment = document.createDocumentFragment()
    fragment.appendChild(document.createTextNode(this.factor))
    fragment.appendChild(document.createTextNode(" "))
    fragment.appendChild(this.unit.toDOM())
    if (this.factor == 1 && !this.isDimensionless()) {
        fragment.removeChild(fragment.firstChild)
        fragment.removeChild(fragment.firstChild)
    } 
    //fragment.normalize()
    return fragment
}

Pacioli.DimensionedNumber.prototype.mult = function (other) {
    return new Pacioli.DimensionedNumber(this.factor * other.factor, this.unit.mult(other.unit));
}

Pacioli.DimensionedNumber.prototype.div = function (other) {
    return new Pacioli.DimensionedNumber(this.factor / other.factor, this.unit.div(other.unit));
}

Pacioli.DimensionedNumber.prototype.reciprocal = function () {
    return new Pacioli.DimensionedNumber(1 / this.factor, this.unit.reciprocal());
}

Pacioli.DimensionedNumber.prototype.expt = function (power) {
    return new Pacioli.DimensionedNumber(Math.pow(this.factor, power), this.unit.expt(power));
}

Pacioli.DimensionedNumber.prototype.map = function (fun) {
    throw "is map on dimensioned numbers used?"
    return new Pacioli.DimensionedNumber(this.factor, this.unit.map(fun));
}

Pacioli.DimensionedNumber.prototype.symbolized = function () {
    throw "is symbolized on dimensioned numbers used?"
    return this.factor + " " + this.unit.symbolized();
}

Pacioli.DimensionedNumber.prototype.conversionFactor = function (to) {
    throw "is conversionFactor on dimensioned numbers used?"
}
